### PR TITLE
Carrier field in invoice must not be displayed if order is virtual

### LINKED ISSUE #20636: Carrier field in invoice must not be displayed if order contains only virtuals products
#### Describe the bug
The Carrier field in the invoice must not be displayed if an order is virtual i.e if all products of an order are virtual.
https://prnt.sc/u1sfdh

#### Expected behavior

The carrier string must not be displayed in the invoice.

#### Steps to Reproduce

1. Go to the front end and add one or more than one virtual product in the cart.
2. Make a successful order.
3. Now go to the back-office and download invoice for that order.
4. Since the order contains all virtual products that mean no shipping is required but the string is appearing
https://prnt.sc/u1sfdh that is incorrect.

**Screenshots**

Ihttps://prnt.sc/u1sfdh

#### Additional information

* PrestaShop version: 1.7.6.7
* PHP version: 7.0

### PR DESCRIPTION
**Pull request by Amit-Kumar-Tiwari-Webkul**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | The Carrier field in the invoice must not be displayed if an order is virtual i.e if all products of an order are virtual.
https://prnt.sc/u1sfdh
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #20636 
| How to test?  | Go to the front end and add one or more than one virtual product in the cart. Make a successful order. Now go to the back-office and download invoice for that order. Since the order contains all virtual products that mean no shipping is required but the string is appearing https://prnt.sc/u1sfdh that is incorrect.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/20637)

### CODE REVIEW

**Review by NeOMakinG — APPROVED** [review]
My bad

File: pdf/invoice.tpl [hunk 1]
```diff
@@ -106,6 +107,7 @@
 		</td>
 		<td colspan="1">&nbsp;</td>
 	</tr>
+{/if}
```
Comment (NeOMakinG) [inline pdf/invoice.tpl]: Hello, thanks for your contribution!

Need some indentation there I think

**Review by NeOMakinG — CHANGES_REQUESTED** [review]

File: pdf/invoice.tpl [hunk 2]
```diff
@@ -98,6 +98,7 @@
 		<td colspan="1">&nbsp;</td>
 	</tr>
 
+{if isset($is_virtual_order) && $is_virtual_order}
```
Comment (PierreRambaud) [inline pdf/invoice.tpl]:
```suggestion
  {if !empty($is_virtual_order)}
```
And you need to increase the indentation of the if content.

**Review by matthieu-rolland — COMMENTED** [review]
You have some feedback on indentation, otherwise your fix looks good to me 👍

**Review by NeOMakinG — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

File: pdf/invoice.tpl [hunk 3]
```diff
@@ -98,14 +98,16 @@
 		<td colspan="1">&nbsp;</td>
 	</tr>
 
-	<tr>
-		<td colspan="6" class="left">
+  {if isset($is_virtual_order) && $is_virtual_order}
+  <tr>
+    <td colspan="6" class="left">
```
Comment (PierreRambaud) [inline pdf/invoice.tpl]:
```suggestion
  {if isset($is_virtual_order) && $is_virtual_order}
    <tr>
      <td colspan="6" class="left">
```

File: pdf/invoice.tpl [hunk 4]
```diff
@@ -98,14 +98,16 @@
...
 
-			{$shipping_tab}
+      {$shipping_tab}
 
-		</td>
-		<td colspan="1">&nbsp;</td>
-	</tr>
+    </td>
+    <td colspan="1">&nbsp;</td>
+  </tr>
```
Comment (PierreRambaud) [inline pdf/invoice.tpl:109]:
```suggestion
      </td>
      <td colspan="1">&nbsp;</td>
    </tr>
```

File: pdf/invoice.tpl [hunk 5]
```diff
@@ -98,14 +98,16 @@
...
 	</tr>
 
-	<tr>
-		<td colspan="6" class="left">
+  {if isset($is_virtual_order) && $is_virtual_order}
+  <tr>
+    <td colspan="6" class="left">
 
-			{$shipping_tab}
+      {$shipping_tab}
```
Comment (PierreRambaud) [inline pdf/invoice.tpl:105]:
```suggestion
        {$shipping_tab}
```

**Review by PierreRambaud — DISMISSED** [review]
I added comments, but the indentation is broken everywhere in this file. So it's ok for me :+1:

**Comment by Amit-Kumar-Tiwari-Webkul** [comment]
HI, Any  Update from QA team

**Comment by khouloudbelguith** [comment]
Hi @Amit-Kumar-Tiwari-Webkul,

I found some issues:

Steps to reproduce the issue:
1. add a virtual product to the cart
2. checkout
3. go to BO > Orders > last order created
4. generate Invoice

=>The Invoice is NOK
Before, I have this
image

After **with PR**
image

I found another issue
Steps to reproduce the issue:
1. add a virtual product to the cart, add another standard product to the cart
2. checkout
3. go to BO > Orders > last order created
4. generate Invoice

=> see the error, the carrier is not displayed => it shouldn't
image
PS: same issue if you have a standard product only => the carrier is not displayed anymore.

Thanks!

**Comment by Amit-Kumar-Tiwari-Webkul** [comment]
@khouloudbelguith , sorry! It's my bad. I am changing some code for fixes.
But the payment method (https://prnt.sc/ua21qi ) is not disappearing from our side testing because there was no change in the payment method code. I have tested it on the ps version1.7.6.7.

**Comment by Amit-Kumar-Tiwari-Webkul** [comment]
@khouloudbelguith ,
After fixes, I have tested both cases, In both cases PR is working
**Case1 :** 1) add a virtual product to the cart
2) checkout
3_ go to BO > Orders > last order created
4) generate Invoice
Result: https://prnt.sc/ua2e7n  -> OK
**case-2:** 
1 ) add a virtual product to the cart, 
2)  add another standard product to the cart
3) checkout
4) go to BO > Orders > last order created
5) generate Invoice
Result : https://prnt.sc/ua2ffc -> OK

**Comment by Amit-Kumar-Tiwari-Webkul** [comment]
@khouloudbelguith, Hi any update on this PR?

**Comment by khouloudbelguith** [comment]
HI @Amit-Kumar-Tiwari-Webkul,

This PR should be reviewed by our developers first :slightly_smiling_face: 

Thanks!

**Review by atomiix — APPROVED** [review]
I'm approving this PR but it would be nice if you could [squash your commits](https://www.internalpointers.com/post/squash-commits-into-one-git) ;)

**Review by Progi1984 — APPROVED** [review]

**Comment by khouloudbelguith** [comment]
Hi @Amit-Kumar-Tiwari-Webkul,

I found an issue.
Steps to reproduce the issue:
1. add a standard product to the cart
2. checkout
3. go to BO > Orders > last order created
4. generate Invoice

=> see the error, the carrier is not well displayed => it shouldn't
image

Before
image

Thanks!

**Comment by Progi1984** [comment]
@Amit-Kumar-Tiwari-Webkul Coud you fix error (given by @khouloudbelguith) ? Thanks

**Comment by Amit-Kumar-Tiwari-Webkul** [comment]
@Progi1984 , I am unable to produce this issue (given by @khouloudbelguith ).
I have tested it on ps1.7.6.8 and ps 1.7.7.0 and working fine https://prnt.sc/ua2ffc
also, I just added a condition on line https://github.com/PrestaShop/PrestaShop/blob/a007e16bc60b87dc0a174224ae9e7ebb59d9352b/pdf/invoice.tpl#L101
do not know how it may cause a design issue?

**Comment by khouloudbelguith** [comment]
Hi @Amit-Kumar-Tiwari-Webkul,

In your screenshot, you have two products ordered (standard product + virtual product).
Thanks to check only with one standard product to see if you can reproduce the issue.

Thanks!

**Comment by Amit-Kumar-Tiwari-Webkul** [comment]
@khouloudbelguith , I have tested it with one standard product.
1) If an order contains virtual products => https://prnt.sc/vdwkps (OK)
2) If an order contains standard products => https://prnt.sc/vdwjeu (OK)
but did not get any issue.
Thanks,

**Comment by khouloudbelguith** [comment]
Hi @Amit-Kumar-Tiwari-Webkul,

I tried again, the same issue.
Steps to reproduce:

1. Fresh Install PS English language/ France country
2. Upload an Indian pack
3. Go to FO > Add demo_8 to the cart (select ₹ as a currency)
4. Continue checkout
5. Select a United states country (no taxes)
6. Place the Order
7. Change the status of the order to Payment accepted
8. Download the invoice
9. See error
image

Thanks!

**Comment by PierreRambaud** [comment]
Hi @Amit-Kumar-Tiwari-Webkul ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**